### PR TITLE
[SPARK-43309][SPARK-38461][CORE] Extend INTERNAL_ERROR with categories and add error class INTERNAL_ERROR_BROADCAST

### PR DESCRIPTION
**Pull request by bozhang2820**
### What changes were proposed in this pull request?

This change is to extend INTERNAL_ERROR with categories, add INTERNAL_ERROR_BROADCAST as an example of it, and change exceptions created in the broadcast package to use that error class.

### Why are the changes needed?

This is to better differentiate internal errors from different categories / areas / modules.

### Does this PR introduce _any_ user-facing change?

Yes. The exceptions created in the broadcast package will change to be of error class INTERNAL_ERROR_BROADCAST.

### How was this patch tested?

Updated unit tests.

### CODE REVIEW

**Comment by bozhang2820** [comment]
The test failure in GitHub Actions "track allocated resources by taskId" seems to be unrelated to this change.

**Comment by bozhang2820** [comment]
@cloud-fan, @srielau, could you take a look at this?

File: core/src/test/scala/org/apache/spark/broadcast/BroadcastSuite.scala [hunk 1]
```diff
@@ -315,9 +315,15 @@ class BroadcastSuite extends SparkFunSuite with LocalSparkContext with Encryptio
     if (removeFromDriver) {
       // Using this variable on the executors crashes them, which hangs the test.
       // Instead, crash the driver by directly accessing the broadcast value.
-      intercept[SparkException] { broadcast.value }
-      intercept[SparkException] { broadcast.unpersist(blocking = true) }
-      intercept[SparkException] { broadcast.destroy(blocking = true) }
+      val e1 = intercept[SparkException] { broadcast.value }
```
Comment (cloud-fan) [inline core/src/test/scala/org/apache/spark/broadcast/BroadcastSuite.scala:318]: can we use the `checkError` util for testing?

Comment (cloud-fan) [reply core/src/test/scala/org/apache/spark/broadcast/BroadcastSuite.scala:318]: nvm, we don't need to check error message here.

**Review by cloud-fan — APPROVED** [review]

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Review by MaxGekk — COMMENTED** [review]
Could you explain, please, why didn't you introduce sub-classes? Like `INTERNAL_ERROR.BROADCAST`

**Comment by cloud-fan** [comment]
@MaxGekk sub error class can't have dedicated sql state.

**Comment by MaxGekk** [comment]
> sub error class can't have dedicated sql state.

ok, but `INTERNAL_ERROR_BROADCAST` and `INTERNAL_ERROR` have the same sql state in the PR. Just in case, can't we override the sql state in sub-classes (are there any principal issues of implementing this?).

**Comment by bozhang2820** [comment]
Another advantage is that having `INTERNAL_ERROR_BROADCAST` will make it possible to have subclasses for the broadcast package.

**Pull request closed without merging**